jsk_perception/src/template_match_detector.cpp

```diff
@@ -0,0 +1,478 @@
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/o2r other materials provided
+ *     with the distribution.
+ *   * Neither the name of the JSK Lab nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+
+#include "jsk_perception/template_match_detector.h"
+#include <boost/assign.hpp>
+#include <jsk_topic_tools/log_utils.h>
+#include <sensor_msgs/image_encodings.h>
+#include <cv_bridge/cv_bridge.h>
+
+namespace jsk_perception
+{
+  void TemplateMatchDetector::onInit()
+  {
+    DiagnosticNodelet::onInit();
+    srv_ = boost::make_shared <dynamic_reconfigure::Server<Config> > (*pnh_);
+    dynamic_reconfigure::Server<Config>::CallbackType f =
+      boost::bind (&TemplateMatchDetector::configCallback, this, _1, _2);
+    srv_->setCallback (f);
+
+    img_pub_ = advertise<sensor_msgs::Image>(*pnh_, "output/viz", 1);
+    rects_pub_ = advertise<jsk_recognition_msgs::RectArray>(*pnh_, "output/rect", 1);
+
+    stored_thre = 0.0;
+#ifdef USE_CUDA
+    if(cv::cuda::getCudaEnabledDeviceCount() > 0){
+      use_cuda = true;
+      cv::cuda::DeviceInfo info(0);
+      ROS_INFO("GPU FOUND : %s", info.name());
+    } else {
+      use_cuda = false;
+      ROS_INFO("NO GPU FOUND : RUN CPU MODE");
+    }
+#else
+      use_cuda = false;
+      ROS_INFO("RUN CPU MODE");
+#endif
+
+    onInitPostProcess();
+  }
+
+  void TemplateMatchDetector::subscribe()
+  {
+    sub_image_.subscribe(*pnh_, "input", 1);
+    sub_info_.subscribe(*pnh_, "input/info", 1);
+    async_ = boost::make_shared<message_filters::Synchronizer<ApproximateSyncPolicy> >(100);
+    async_->connectInput(sub_image_, sub_info_);
+    async_->registerCallback(boost::bind(&TemplateMatchDetector::apply, this, _1, _2));
```

Comment:
Could you add `~approximate_sync: true/false` and `~queue_size`?

Reply:
add